File: test/data/cpp/test04/StateNotShooting.cpp
#include "StateNotShooting.h"
StateNotShooting::StateNotShooting(StateController* stateController, IControllee* controllee) : BaseState(controllee)
{
    _stateController = stateController;
    _controllee = controllee;
}
StateNotShooting::~StateNotShooting()
{
}
void StateNotShooting::Setup(bool resume, bool deepResume)
{
    if(!deepResume)
    {
        SetupSubState(_stateController->InstanceOfIdle, resume);
    }
}
BaseState* StateNotShooting::TransitEscape()
{
    _controllee->DoAction3();
    _stateController->InstanceOfEscaped->Setup(false, false);
    return _stateController->InstanceOfEscaped;
}
BaseState* StateNotShooting::TransitEvConfig()
{
    BaseState* currentSubState = CurrentSubState();
    if(currentSubState != nullptr)
    {
        BaseState* nextState = currentSubState->TransitEvConfig();
        return TransitBySubState(nextState);
    }
    return nullptr;
}
const char* StateNotShooting::GetStateName()
{
    BaseState* currentSubState = CurrentSubState();
    if(currentSubState == nullptr)
    {
        return "NotShooting(end)";
    }
    return currentSubState->GetStateName();
}
BaseState* StateNotShooting::GetParent()
{
    return nullptr;
}
int StateNotShooting::GetStateID()
{
    return 4;
}

File: test/data/cs/test04/StateNotShooting.cs
public class StateNotShooting : BaseState
{
    private StateController _stateController;
    private IControllee _controllee;
    public StateNotShooting(StateController stateController, IControllee controllee) : base(controllee)
    {
        _stateController = stateController;
        _controllee = controllee;
    }
    public override void Setup()
    {
        SetupSubState(_stateController.InstanceOfIdle);
    }
    public override BaseState? TransitEscape()
    {
        _controllee.DoAction3();
        _stateController.InstanceOfEscaped.Setup();
        return _stateController.InstanceOfEscaped;
    }
    public override BaseState? TransitEvConfig()
    {
        BaseState? currentSubState = CurrentSubState();
        if(currentSubState != null)
        {
            BaseState? nextState = currentSubState.TransitEvConfig();
            return TransitBySubState(nextState);
        }
        return null;
    }
    public override string GetStateName()
    {
        BaseState? currentSubState = CurrentSubState();
        if(currentSubState == null)
        {
            return "NotShooting(end)";
        }
        return currentSubState.GetStateName();
    }
    public override BaseState? GetParent()
    {
        return null;
    }
}